lib/redis/semaphore.rb

```diff
@@ -19,13 +19,17 @@ def initialize(name, opts = {})
       @expiration = opts.delete(:expiration)
       @resource_count = opts.delete(:resources) || 1
       @stale_client_timeout = opts.delete(:stale_client_timeout)
-      @redis = opts.delete(:redis) || Redis.new(opts)
       @use_local_time = opts.delete(:use_local_time)
+      @redis = if @redis = opts.delete(:redis)
+                 @redis.dup
+               else
+                 Redis.new(opts)
+               end
       @tokens = []
     end
 
     def exists_or_create!
-      token = @redis.getset(exists_key, EXISTS_TOKEN)
```

Comment:
Removing `#getset` destroys the atomicity of initialization. We want to keep `getset`.